deku-core/src/Deku/Core.purs

```diff
@@ -232,316 +157,170 @@ unsafeAttribute = Attribute
-       | r
-       }
-  -> Maybe
-       { e :: DekuBeacon
-       , l :: Poll DekuDynamic
-       , s :: DekuBeacon
-       , lucky :: Ref Boolean
-       }
-getLifecycle mb = do
-  m <- mb
-  l <- m.lifecycle
-  pure { l, s: m.start, e: m.end, lucky: m.lucky }
-
-thunker :: STArray.STArray Global (Effect Unit) -> Effect Unit
-thunker unsubs = do
-  unsubsX <- liftST $ STArray.unsafeFreeze unsubs
-  runEffectFn1 fastForeachThunkE unsubsX
-
-runListener
+collectUnsubs :: EffectFn1 PSR (STArray.STArray Global (Effect Unit))
+collectUnsubs = mkEffectFn1 \(PSR psr) -> do
+  unsubs <- liftST $ STArray.new
+  when (not (Array.null psr.unsubs)) do
+    void $ liftST $ STArray.pushAll psr.unsubs unsubs
+  pure unsubs
+
+disposeUnsubs :: EffectFn1 (STArray.STArray Global (Effect Unit)) Unit
+disposeUnsubs = mkEffectFn1 \unsubs -> do
+  runEffectFn1 Event.fastForeachThunkE =<< liftST (STArray.unsafeFreeze unsubs)
+
+-- | Handles an optimized `Poll` by running the effect on each emitted value. Any resulting subscription gets written to 
+-- | the given cleanup array.
+pump
   :: forall a
-   . EffectFn1 a Unit
-  -> STArray.STArray Global (Effect Unit)
+   . STArray.STArray Global (Effect Unit)
   -> Poll a
+  -> EffectFn1 a Unit
   -> Effect Unit
-runListener oh'hi associations = go
+pump associations p eff =
+  go p
+
   where
+
+  handleEvent :: Event.Event a -> Effect Unit
   handleEvent y = do
-    uu <- runEffectFn2 subscribeO y oh'hi
+    uu <- runEffectFn2 Event.subscribeO y eff
     void $ liftST $ STArray.push uu associations
+
+  go :: Poll a -> Effect Unit
   go = case _ of
     OnlyEvent x -> handleEvent x
-    OnlyPure x -> runEffectFn2 fastForeachE x oh'hi
+    OnlyPure x -> runEffectFn2 Event.fastForeachE x eff
     OnlyPoll x -> do
-      pump <- liftST $ Event.create
-      handleEvent (UPoll.sample x pump.event)
-      pump.push identity
+      bang <- liftST $ Event.create
+      handleEvent (UPoll.sample x bang.event)
+      bang.push identity
     PureAndEvent x y -> do
       go (OnlyPure x)
       go (OnlyEvent y)
     PureAndPoll x y -> do
       go (OnlyPure x)
       go (OnlyPoll y)
 
-notLucky :: Ref Boolean -> Effect Unit
-notLucky = write false
+newtype PSR = PSR
+  { lifecycle :: Poll.Poll Unit
+  -- used by `Nut`s to register or clear the last element of their region.
+  , region :: StaticRegion
+  , unsubs :: Array (Effect Unit)
+  }
+
+derive instance Newtype PSR _
+
+newtype Nut =
+  Nut (EffectFn2 PSR DOMInterpret Unit)
+
+instance Semigroup Nut where
+  append (Nut a) (Nut b) =
+    -- unrolled version of `fixed`
+    Nut $ mkEffectFn2 \psr di -> do
+      -- first `Nut` should not handle any unsubs, they may still be needed for later elements
+      runEffectFn2 a (over PSR _ { unsubs = [] } psr) di
+      runEffectFn2 b psr di
+
+instance Monoid Nut where
+  mempty =
+    -- while we contribute no UI elements we still have to handle any unsubs created by our hooks
+    Nut $ mkEffectFn2 \psr _ -> do
```

Comment:
It's gonna be tough for me to do a thorough review of the PR due to its size, and I'm sure I'll merge it, so these are mostly questions for reflection.

Usually, `mempty` is pretty vanilla, meaning we don't count on it to do anything. Here, it triggers side effects at some point, and elsewhere in the code, it seems like it's used explicitly to trigger those side effects, ie `Nut dispose = mempty`.

I get that an empty nut may still need to trigger some effects in case it's the receptacle of a bunch of side effects that reside nowhere else, but still, something about this seems like the sort of thing that's worth a think. That said, I can't offer a more productive comment as I'm still not well-versed enough in the code yet, so it's just food for thought!

Reply:
Well it's a matter of definition I guess. If we define this aspect of `Nut` as "a scoped block that guarantees that all deferred effects get run once when `lifecycle` has a signal" then even the empty element has to do that. And `append` in turn also honors this contract.
Although it's probably better to factor out this simple lifecycle handling to make the code in `fixed` more clear.